astropy/io/registry/registry.py

```diff
@@ -0,0 +1,589 @@
+            else:
+                raise TypeError("reader should return a {0} instance".format(cls.__name__))
+    finally:
+        if ctx is not None:
+            ctx.__exit__(*sys.exc_info())
+
+    return data
+
+
+def write(data, *args, **kwargs):
+    """
+    Write out data
+
+    The arguments passed to this method depend on the format
+    """
+
+    if 'format' in kwargs:
+        format = kwargs.pop('format')
+    else:
+        format = None
+
+    if format is None:
+        path = None
+        fileobj = None
+        if len(args):
+            if isinstance(args[0], six.string_types):
+                path = args[0]
+                fileobj = None
+            elif hasattr(args[0], 'read'):
+                path = None
+                fileobj = args[0]
+
+        format = _get_valid_format(
+            'write', data.__class__, path, fileobj, args, kwargs)
+
+    writer = get_writer(format, data.__class__)
+    writer(data, *args, **kwargs)
+
+
+def _get_valid_format(mode, cls, path, fileobj, args, kwargs):
+    """
+    Returns the first valid format that can be used to read/write the data in
+    question.  Mode can be either 'read' or 'write'.
+    """
+
+    valid_formats = identify_format(mode, cls, path, fileobj, args, kwargs)
+
+    if len(valid_formats) == 0:
+        format_table_str = _get_format_table_str(cls, mode.capitalize())
+        raise Exception("Format could not be identified.\n"
+                        "The available formats are:\n"
+                        "{0}".format(format_table_str))
+    elif len(valid_formats) > 1:
+        raise Exception(
+            "Format is ambiguous - options are: {0}".format(
+                ', '.join(sorted(valid_formats))))
+
+    return valid_formats[0]
+
+
+class MetaRegisterBaseIO(type):
+
+    def __init__(cls, name, bases, members):
+
+        super(MetaRegisterBaseIO, cls).__init__(name, bases, members)
+
+        format_abbreviation = members.get('_format_name')
+        if format_abbreviation is None:
+            if cls.__name__ == 'BaseIO':
+                return
+            else:
+                raise ValueError("_format_name is not defined")
+
+        supported_class = members.get('_supported_class')
+        if supported_class is None:
+            raise ValueError("_supported_class is not defined")
+
+        import inspect
+        frm = inspect.stack()[1]
+        module = inspect.getmodule(frm[0])
+
+        resolve_data_classes()
+
+        if not supported_class in _registry:
+            _registry[supported_class] = {}
+
+        _registry[supported_class][format_abbreviation] = {
+            'module': module,
+            'io_class': cls,
+            'read': members.get('read') is not None,
+            'write':  members.get('write') is not None,
+            'identify':  members.get('identify') is not None
+        }
+
+
+@six.add_metaclass(MetaRegisterBaseIO)
+class BaseIO(object):
+
+    _format_name = None
```

Comment:
Kind of a trivial point, but I wonder if these should be considered public attributes.  For the `BaseIO` class and subclasses, these attributes feel like part of the key public functionality.  (Though I don't understand what, if anything, the `_flexible` attribute does since it doesn't seem to be used anywhere).